### PR TITLE
encryption-at-rest: Update

### PR DESCRIPTION
**Pull request by dveeden**
### What is changed, added or deleted? (Required)

Updating "Encryption at Rest" docs:

- Add note about using AWS KMS for on-premise deployment
- Add notes about key creation (e.g. to use the right key type)
- Fix the syntax of the encryption section configuration
- Add note about VPC endpoints for KMS
- Add note about multi-Region keys
- Add information about data-at-rest support in PD ( https://github.com/pingcap/tidb/issues/18262 )
- Remove some of the references to pre-v5.0 versions ( https://github.com/pingcap/docs/pull/6152#issuecomment-898099044 )
- Use `168h` instead of `7d` as PD doesn't support `d` as unit specifier. ( https://github.com/tikv/pd/issues/3987 )
- Added a note about EBS Encryption

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### CODE REVIEW

**Comment by dveeden** [comment]
@TomShawn should we remove all reference to specific 4.x versions in master? I think that makes it easier to read etc.

**Comment by dveeden** [comment]
This suggests PD might support TDE, but I can't find anything about that in the docs.
```
mysql> show config where Name like 'security.encryption.%';
+------+-----------------+-------------------------------------------------------+--------------------------------------+
| Type | Instance        | Name                                                  | Value                                |
+------+-----------------+-------------------------------------------------------+--------------------------------------+
| pd   | 127.0.0.1:2379  | security.encryption.data-encryption-method            | plaintext                            |
| pd   | 127.0.0.1:2379  | security.encryption.data-key-rotation-period          | 168h0m0s                             |
| pd   | 127.0.0.1:2379  | security.encryption.master-key.endpoint               |                                      |
| pd   | 127.0.0.1:2379  | security.encryption.master-key.key-id                 |                                      |
| pd   | 127.0.0.1:2379  | security.encryption.master-key.path                   |                                      |
| pd   | 127.0.0.1:2379  | security.encryption.master-key.region                 |                                      |
| pd   | 127.0.0.1:2379  | security.encryption.master-key.type                   | plaintext                            |
| tikv | 127.0.0.1:20160 | security.encryption.data-encryption-method            | aes256-ctr                           |
| tikv | 127.0.0.1:20160 | security.encryption.data-key-rotation-period          | 1d                                   |
| tikv | 127.0.0.1:20160 | security.encryption.enable-file-dictionary-log        | true                                 |
| tikv | 127.0.0.1:20160 | security.encryption.file-dictionary-rewrite-threshold | 1000000                              |
| tikv | 127.0.0.1:20160 | security.encryption.master-key.endpoint               |                                      |
| tikv | 127.0.0.1:20160 | security.encryption.master-key.key-id                 | e857425b-6a5f-41df-a56f-6bcdb0fffd14 |
| tikv | 127.0.0.1:20160 | security.encryption.master-key.region                 | eu-central-1                         |
| tikv | 127.0.0.1:20160 | security.encryption.master-key.type                   | kms                                  |
| tikv | 127.0.0.1:20160 | security.encryption.master-key.vendor                 |                                      |
| tikv | 127.0.0.1:20160 | security.encryption.previous-master-key.type          | plaintext                            |
+------+-----------------+-------------------------------------------------------+--------------------------------------+
17 rows in set (0.01 sec)
```

This seems to be linked to https://github.com/pingcap/tidb/issues/18262.

@yiwu-arbug any reason why this should not be in the docs? Maybe with a note about this being experimental?

**Comment by TomShawn** [comment]
> @TomShawn should we remove all reference to specific 4.x versions in master? I think that makes it easier to read etc.

Yes. Of course.

File: encryption-at-rest.md [hunk 1]
```diff
@@ -4,20 +4,24 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+TiKV supports encryption at rest. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest.
 
-Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encryption.
+TiFlash also supports encryption at rest and while it only stores some metadata PD also supports data at rest. Encryption at rest can be enabled per component.
```
Comment (yiwu-arbug) [inline encryption-at-rest.md:13]: Please note that PD's encryption-at-rest support is experimental.

**Review by yiwu-arbug — COMMENTED** [review]
rest LGTM

File: encryption-at-rest.md (hunk 1 above)
Comment (dveeden) [reply encryption-at-rest.md:17]: I've added a warning for that

File: encryption-at-rest.md [hunk 2]
```diff
@@ -4,20 +4,32 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
 
-* When a TiDB cluster is deployed, the majority of user data is stored in TiKV nodes, and that data will be encrypted when encryption is enabled. However, a small amount of user data is stored in PD nodes as metadata (for example, secondary index keys used as TiKV region boundaries). As of v4.0.0, PD doesn't support encryption at rest. It is recommended to use storage-level encryption (for example, file system encryption) to help protect sensitive data stored in PD.
-* TiFlash supports encryption at rest since v4.0.5. For details, refer to [Encryption at Rest for TiFlash](#encryption-at-rest-for-tiflash-new-in-v405). When deploying TiKV with TiFlash earlier than v4.0.5, data stored in TiFlash is not encrypted.
+* When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes, and that data will be encrypted when encryption is enabled. However, a small amount of user data is stored in PD nodes as metadata (for example, secondary index keys used as TiKV region boundaries). PD also supports encryption at rest.
+* TiFlash supports encryption at rest, for details, refer to [Encryption at Rest for TiFlash](#encryption-at-rest-for-tiflash).
```
Comment (yiwu-arbug) [inline encryption-at-rest.md:32]: I think we can refactor this a bit. We can remove these two warning items, but add a dedicated paragraph somewhere noting encryption should be enable per component. We can merge the warning of PD encryption being experimental here.

Comment (dveeden) [reply encryption-at-rest.md:32]: Would it be ok to merge this PR and then let you take care of refactoring it? Or could you add a suggestion?

File: encryption-at-rest.md [hunk 3]
```diff
@@ -4,23 +4,45 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+## Per component encryption support
+
+When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes. Some metadata is stored in PD nodes (for example, secondary index keys used as TiKV region boundaries). To get the full benefits of this feature encryption needs to be enabled for all components. Backups, log files and data transmitted over the network should also be considered when implementing this.
```
Comment (yiwu-arbug) [inline encryption-at-rest.md:17]:
```suggestion
When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes. Some metadata is stored in PD nodes (for example, secondary index keys used as TiKV region boundaries). To get the full benefits of this feature encryption needs to be enabled for all components. Backups and data transmitted over the network (TLS) should also be considered when implementing this.
```
We don't have feature to encrypt the log files, so probably don't mention it.

File: encryption-at-rest.md [hunk 4]
```diff
@@ -4,23 +4,45 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
-* TiFlash supports encryption at rest since v4.0.5. For details, refer to [Encryption at Rest for TiFlash](#encryption-at-rest-for-tiflash-new-in-v405). When deploying TiKV with TiFlash earlier than v4.0.5, data stored in TiFlash is not encrypted.
-* TiKV currently does not exclude encryption keys and user data from core dumps. It is advised to disable core dumps for the TiKV process when using encryption at rest. This is not currently handled by TiKV itself.
-* TiKV tracks encrypted data files using the absolute path of the files. As a result, once encryption is turned on for a TiKV node, the user should not change data file paths configuration such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.
-* TiKV, TiDB, and PD  info logs might contain user data for debugging purposes. The info log and this data in it are not encrypted. It is recommended to enable [log redaction](/log-redaction.md).
```
Comment (yiwu-arbug) [inline encryption-at-rest.md:23]: Let's keep the last three warning items. They are still valid.

File: encryption-at-rest.md (hunk 3 above)
Comment (dveeden) [reply encryption-at-rest.md:17]: But there is https://docs.pingcap.com/tidb/stable/log-redaction and the option to change the log level.

File: encryption-at-rest.md (hunk 4 above)
Comment (dveeden) [reply encryption-at-rest.md:23]: These were not removed, just moved around.

Comment (yiwu-arbug) [reply encryption-at-rest.md:23]: I see.

File: encryption-at-rest.md (hunk 3 above)
Comment (yiwu-arbug) [reply encryption-at-rest.md:17]: got it. specifically mention log redaction and link to the docs? also for TLS.

Comment (dveeden) [reply encryption-at-rest.md:17]: There is a `Logging` section that mentions log redaction and links to it.

As logging is stored on disk (at rest) and TLS is not I decided to not go into much detail on TLS here.

**Review by yiwu-arbug — APPROVED** [review]
LGTM

File: encryption-at-rest.md [hunk 5]
```diff
@@ -4,23 +4,45 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+
+### PD
+
+Using encryption-at-rest for PD is an experimental feature. This is configured in the same way as TiKV.
 
-Also from v4.0.0, BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encryption.
+### Backups with BR
 
-## Warnings
+BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encryption. See [BR S3 server-side encryption](/encryption-at-rest.md#br-s3-server-side-encryption) for details.
```
Comment (TomShawn) [inline encryption-at-rest.md:41]:
```suggestion
BR supports S3 server-side encryption (SSE) when backing up data to S3. A customer-owned AWS KMS key can also be used together with S3 server-side encryption. See [BR S3 server-side encryption](/encryption-at-rest.md#br-s3-server-side-encryption) for details.
```

File: encryption-at-rest.md [hunk 6]
```diff
@@ -39,14 +61,27 @@ Data keys are generated by TiKV and passed to the underlying storage engine (nam
 
 Regardless of data encryption method, data keys are encrypted using AES256 in GCM mode for additional authentication. This required the master key to be 256 bits (32 bytes), when passing from file instead of KMS.
 
+### Key creation
+
+Go to the [AWS KMS](https://console.aws.amazon.com/kms) on the AWS console. Make sure the correct region is selected on the top right corner of your console. Make  and click "Create a key". Select "Symmetric" as Key type. After this you can set an alias an description and set tags.
```
Comment (TomShawn) [inline encryption-at-rest.md:66]:
```suggestion

To create a key on AWS, follow these steps:

1. Go to the [AWS KMS](https://console.aws.amazon.com/kms) on the AWS console.
2. Make sure that you have selected the correct region on the top right corner of your console.
3. Click **Create a key** and select **Symmetric** as the key type.
4. Set an alias for the key and set tags.
```

File: encryption-at-rest.md [hunk 7]
```diff
@@ -39,14 +61,27 @@ Data keys are generated by TiKV and passed to the underlying storage engine (nam
...
+
+It is also possible to do this with the AWS Cli:
+
+```
```
Comment (TomShawn) [inline encryption-at-rest.md:70]:
```suggestion
```shell
```

File: encryption-at-rest.md [hunk 8]
```diff
@@ -61,7 +96,9 @@ region = "us-west-2"
 endpoint = "https://kms.us-west-2.amazonaws.com"
 ```
 
-The `key-id` specifies the key id for the KMS CMK. The `region` is the AWS region name for the KMS CMK. The `endpoint` is optional and doesn't need to be specified normally, unless you are using a AWS KMS compatible service from a non-AWS vendor.
+The `key-id` specifies the key id for the KMS CMK. The `region` is the AWS region name for the KMS CMK. The `endpoint` is optional and doesn't need to be specified normally, unless you are using a AWS KMS compatible service from a non-AWS vendor or need to use a [VPC endpoint for KMS](https://docs.aws.amazon.com/kms/latest/developerguide/kms-vpc-endpoint.html).
+
+It is possible to use [multi-Region keys](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html). For this you need to setup a primary key in a specific region and add replica keys in the regions you require.
```
Comment (TomShawn) [inline encryption-at-rest.md:101]:
```suggestion
You can also use [multi-Region keys](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html) in AWS. For this, you need to set up a primary key in a specific region and add replica keys in the regions you require.
```

File: encryption-at-rest.md [hunk 9]
```diff
@@ -4,23 +4,45 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+
+### TiKV
+
+TiKV supports encryption at rest, this allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest
+
+It is possible to use AWS KMS for on-premise deployments, but this is not required.
```
[the remark this answers is not in the document]
Comment (dveeden) [reply encryption-at-rest.md]: Updated. PTAL

File: encryption-at-rest.md [hunk 10]
```diff
@@ -4,23 +4,45 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+
+When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes. Some metadata is stored in PD nodes (for example, secondary index keys used as TiKV region boundaries). To get the full benefits of this feature encryption needs to be enabled for all components. Backups, log files and data transmitted over the network should also be considered when implementing this.
+
+### TiKV
+
+TiKV supports encryption at rest, this allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest
```
[the remark this answers is not in the document]
Comment (dveeden) [reply encryption-at-rest.md]: Updated. PTAL

**Comment by dveeden** [comment]
@TomShawn PTAL

File: encryption-at-rest.md (hunk 6 above)
Comment (TomShawn) [reply encryption-at-rest.md:68]: @dveeden PTAL

File: encryption-at-rest.md [hunk 11]
```diff
@@ -4,23 +4,47 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+
+When a TiDB cluster is deployed, the majority of user data is stored on TiKV and TiFlash nodes. Some metadata is stored on PD nodes (for example, secondary index keys used as TiKV Region boundaries). To get the full benefits of encryption at rest, you need to enable encryption for all components. Backups, log files, and data transmitted over the network should also be considered when you implement encryption.
+
+### TiKV
+
+TiKV supports encryption at rest, this allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key that is used to encrypt the data keys is provided by the user. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest
```
Comment (TomShawn) [inline encryption-at-rest.md:23]: See the duplicated content.
image

Please make sure the same information is only stated once in this doc if not for emphatic purposes.

File: encryption-at-rest.md [hunk 12]
```diff
@@ -4,23 +4,47 @@ summary: Learn how to enable encryption at rest to protect sensitive data.
...
+
+AWS KMS can be used for both Cloud and on-premise deployments, but this is not required. It is also possible to supply the plaintext master key in a file.
```
Comment (TomShawn) [inline encryption-at-rest.md:25]:
```suggestion
Optionally, you can use AWS KMS for both cloud and on-premises deployments. You can also supply the plaintext master key in a file.
```

File: encryption-at-rest.md (hunk 11 above)
Comment (dveeden) [reply encryption-at-rest.md:23]: Like this? 1a5b3c5e32db6ce6825dc3b97dc3eef697efbcef

Comment (TomShawn) [reply encryption-at-rest.md:23]: LGTM

Comment (TomShawn) [reply encryption-at-rest.md:23]:
```suggestion
TiKV supports encryption at rest. This feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by the user and this key is called master key. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest.
```

**Pull request merged**